task/control-plane-agent/src/mgs_gimlet.rs

```diff
@@ -885,22 +998,49 @@ impl UsartHandler {
             // self.from_rx.len()`, which is where we actually are now).
         }
 
-        // Recv as much as we can from the USART FIFO, even if we have to
-        // discard old data to do so.
         let mut n_received = 0;
         let mut discarded_data = 0;
-        while let Some(b) = self.usart.try_rx_pop() {
-            n_received += 1;
-            match self.from_rx.push_back(b) {
-                Ok(()) => (),
-                Err(b) => {
-                    // If `push_back` failed, we know there is at least one
-                    // item, allowing us to unwrap `pop_front`. At that point we
-                    // know there's space for at least one, allowing us to
-                    // unwrap a subsequent `push_back`.
-                    self.from_rx.pop_front().unwrap_lite();
-                    self.from_rx.push_back(b).unwrap_lite();
-                    discarded_data += 1;
+
+        // If our client is Humility, we only want to read from the usart if we
+        // have room in our buffer (i.e., we want to make use of flow control
+        // because Humility is a _very slow_ reader).
+        //
+        // If our client is MGS (the default), we always recv as much as we can
+        // from the USART FIFO, even if that means discarding old data. If an
+        // MGS instance is actually attached, we should be flushing packets out
+        // to it fast enough that we don't see this in practice.
+        if self.client == UartClient::Humility {
```

Comment:
Nit: use a `match` here just in case someone adds a _third_ client type?

Reply:
Fixed in e0eefae4